features/users/users_controller.rb

```diff
@@ -0,0 +1,73 @@
+require "set"
+
+require_relative "../../shared/authenticated_controller_base"
+require_relative "../../services/services"
+
+module FastlaneCI
+  # A CRUD controller to manage users
+  class UsersController < AuthenticatedControllerBase
+    HOME = "/users"
+
+    get HOME do
+      locals = { title: "Users" }
+      erb(:users, locals: locals, layout: FastlaneCI.default_layout)
+    end
+
+    # When the `/users/create` form is submitted:
+    #
+    # - creates a user if the locals are valid
+    post "#{HOME}/create" do
+      if valid_params?(params, users_params)
+        Services.user_service.create_user!(
+          id: params[:id],
+          email: params[:email],
+          password: params[:password]
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    # Updates a user existing in the configuration repository `users.json`
+    post "#{HOME}/update" do
+      if valid_params?(params, users_params)
+        new_user = User.new(
+          id: params[:id],
+          email: params[:email],
+          password: params[:password]
+        )
+
+        Services.user_service.update_user!(new_user)
+      end
+
+      redirect(HOME)
+    end
+
+    private
+
+    #####################################################
+    # @!group Data: View-specific data
+    #####################################################
+
+    # @return [Array[User]]
+    def users
+      Services.user_service.user_data_source.users
```

Comment:
same comment here about accessing what should be private properties in the `user_service`